### PR TITLE
Kie issues#1448 - matches() function wrongly behaves - negation syntax translation

### LINKED ISSUE #1448: `matches()` function wrongly behaves
According to the DMN specs, the `matches()` and `replace()` functions should behave according to the xQuery 1.0 specification. 
Our current implementation doesn't rely on that, using the native Java implementation to manage both.
Unfortunately, the Java regex management differs from the xQuery specs, and natively implementing the xQuery specs leads to huge effort (and complex) work. For that reason, we agreed to rely on an external library that already implemented xQuery specs in Java. 
After a quick analysis (https://en.wikipedia.org/wiki/XQuery#Implementations), we choose [Saxon-He](https://github.com/Saxonica/Saxon-HE) as the best fit for this purpose. 

The scope of the ticket is to integrate this new external dependency into our code base, so both `matches()` and `replace()` can rely on that implementation to correctly behave.

=== _Original description_  ====

TCK Tests revealed some cases where matches() wrongly behaves.

1.     matches("O", "[A-Z-[OI]]", "i") should return false. Before fixing that, double-check in the specs that this is the expected behavior
2.     matches("i", "[A-Z-[OI]]", "i") should return false. Before fixing that, double-check in the specs that this is the expected behavior
 
 
This syntax is invalid for java Pattern
`[A-Z-[OI]]`
and correct one is
`[A-Z&&[^OI]]` (see **_Character classes_** in [docs](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html))

[XML Schema Part 2: Datatypes Second Edition](https://www.w3.org/TR/xmlschema-2/)
makes a distinction between _negation_ (`ˆ`) and _subtraction_ (`-`)
See "Negative Character Group" and "Character Class Subtraction" at [regex chapter](https://www.w3.org/TR/xmlschema-2/#regexs).

_**A negative character group is a [·positive character group·](https://www.w3.org/TR/xmlschema-2/#dt-poschargroup) preceded by the ^ character. For all [·positive character group·](https://www.w3.org/TR/xmlschema-2/#dt-poschargroup)s P, ^P is a valid negative character group, and C(^P) contains all XML characters that are not in C(P).**_

_**A character class subtraction is a [·character class expression·](https://www.w3.org/TR/xmlschema-2/#dt-charexpr) subtracted from a [·positive character group·](https://www.w3.org/TR/xmlschema-2/#dt-poschargroup) or [·negative character group·](https://www.w3.org/TR/xmlschema-2/#dt-negchargroup), using the - character.**_

A "translation" between the different syntax is required

### PR DESCRIPTION
**Pull request by bncriju**
Fixes https://github.com/apache/incubator-kie-issues/issues/1448

**Thank you for submitting this pull request**

**NOTE!:** Double-check the target branch for this PR.
The default is `main` so it will target Drools 8 / Kogito.

**Ports** If a forward-port or a backport is needed, paste the forward port PR here

* [link](https://www.example.com)

**Issue**: _(please edit the GitHub Issues link if it exists)_

* [link](https://www.example.com)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

**Comment by kie-ci3** [comment]
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-drools -u https://github.com/apache/incubator-kie-drools/pull/6083 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/display/redirect

**Test results:**
- PASSED: 22320
- FAILED: 31

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.documentation/ADocFEELExamplesTest/test/">org.kie.dmn.feel.documentation.ADocFEELExamplesTest.test</a></summary>
Cannot invoke "java.net.URL.toURI()" because "resource" is null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.documentation/FromSpecificationNotInAdocTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[21]/">org.kie.dmn.feel.documentation.FromSpecificationNotInAdocTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[21]</a></summary>
[Evaluating: 'matches("foobar", "^fo*b") = true'] <br/>expected: true<br/> but was: false
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.documentation/FromSpecificationNotInAdocTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[22]/">org.kie.dmn.feel.documentation.FromSpecificationNotInAdocTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[22]</a></summary>
[Evaluating: 'matches("foobar", "^fo*b") = true'] <br/>expected: true<br/> but was: false
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[3]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[3]</a></summary>
[Evaluating: 'replace( "  foo   bar zed  ", "^(\s)+|(\s)+$|\s+(?=\s)", "" )'] <br/>expected: "foo bar zed"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[4]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[4]</a></summary>
[Evaluating: 'replace( "  foo   bar zed  ", "^(\s)+|(\s)+$|\s+(?=\s)", "" )'] <br/>expected: "foo bar zed"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[71]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[71]</a></summary>
[Evaluating: 'matches("foo", "[a-z]{3}")'] <br/>expected: true<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[72]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[72]</a></summary>
[Evaluating: 'matches("foo", "[a-z]{3}")'] <br/>expected: true<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[73]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[73]</a></summary>
[Evaluating: 'matches("banana", "[a-z]{3}")'] <br/>expected: true<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[74]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[74]</a></summary>
[Evaluating: 'matches("banana", "[a-z]{3}")'] <br/>expected: true<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[75]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[75]</a></summary>
[Evaluating: 'matches("two \n lines", "two.*lines")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[76]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[76]</a></summary>
[Evaluating: 'matches("two \n lines", "two.*lines")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[79]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[79]</a></summary>
[Evaluating: 'matches("one\ntwo\nthree", "^two$")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[80]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[80]</a></summary>
[Evaluating: 'matches("one\ntwo\nthree", "^two$")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[83]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[83]</a></summary>
[Evaluating: 'matches("FoO", "foo")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[84]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[84]</a></summary>
[Evaluating: 'matches("FoO", "foo")'] <br/>expected: false<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[87]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[87]</a></summary>
[Evaluating: 'replace("banana","a","o")'] <br/>expected: "bonono"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[88]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[88]</a></summary>
[Evaluating: 'replace("banana","a","o")'] <br/>expected: "bonono"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[89]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[89]</a></summary>
[Evaluating: 'replace("banana","(an)+", "**")'] <br/>expected: "b**a"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[90]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[90]</a></summary>
[Evaluating: 'replace("banana","(an)+", "**")'] <br/>expected: "b**a"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[91]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[91]</a></summary>
[Evaluating: 'replace("banana","[aeiouy]","[$0]")'] <br/>expected: "b[a]n[a]n[a]"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[92]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[92]</a></summary>
[Evaluating: 'replace("banana","[aeiouy]","[$0]")'] <br/>expected: "b[a]n[a]n[a]"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[93]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[93]</a></summary>
[Evaluating: 'replace("0123456789","(\d{3})(\d{3})(\d{4})","($1) $2-$3")'] <br/>expected: "(012) 345-6789"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime/FEELFunctionsTest/instanceTest(String,_Object,_Severity,_FEEL_TARGET,_Boolean,_FEELDialect)[94]/">org.kie.dmn.feel.runtime.FEELFunctionsTest.instanceTest(String, Object, Severity, FEEL_TARGET, Boolean, FEELDialect)[94]</a></summary>
[Evaluating: 'replace("0123456789","(\d{3})(\d{3})(\d{4})","($1) $2-$3")'] <br/>expected: "(012) 345-6789"<br/> but was: null
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/RangeFunctionTest/evaluateWithInvalidFunctionInvocationNode/">org.kie.dmn.feel.runtime.functions.RangeFunctionTest.evaluateWithInvalidFunctionInvocationNode</a></summary>
Expected 'retrieved.isLeft()' from, [matches("foo", "[a-z]{3}")..matches("foo", "[a-z]{3}")]
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithoutFlagsPatternNotMatches/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithoutFlagsPatternNotMatches</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.BMPString to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithFlagDotAll/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithFlagDotAll</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.BMPString to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithFlagMultiline/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithFlagMultiline</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.BMPString to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithAllFlags/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithAllFlags</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.BMPString to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithFlagCaseInsensitive/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithFlagCaseInsensitive</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.Twine16 to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.runtime.functions/ReplaceFunctionTest/invokeWithoutFlagsPatternMatches/">org.kie.dmn.feel.runtime.functions.ReplaceFunctionTest.invokeWithoutFlagsPatternMatches</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.Twine16 to java.lang.String
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/drools/job/main/job/pullrequest_jobs/job/drools-pr/job/PR-6083/1/testReport/org.kie.dmn.feel.util/XQueryImplUtilTest/executeReplaceFunctionTest/">org.kie.dmn.feel.util.XQueryImplUtilTest.executeReplaceFunctionTest</a></summary>
java.lang.ClassCastException: Cannot cast net.sf.saxon.str.Twine16 to java.lang.String
</details>

File: kie-dmn/kie-dmn-feel/pom.xml [hunk 1]
```diff
@@ -135,6 +149,10 @@
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-databind</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.poi</groupId>
+      <artifactId>poi</artifactId>
+    </dependency>
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/pom.xml:155]: @bncriju I guess this is not required

File: kie-dmn/kie-dmn-feel/pom.xml [hunk 2]
```diff
@@ -144,7 +162,7 @@
           <directory>${project.basedir}/src/test/resources</directory>
       </testResource>
       <testResource>
-        <directory>${project.basedir}/</directory>
+        <directory>${project.basedir}/../</directory>
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/pom.xml:165]: @bncriju I guess this change is not required

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/MatchesFunction.java [hunk 3]
```diff
@@ -39,21 +36,24 @@ private MatchesFunction() {
         super( "matches" );
     }
 
-    public FEELFnResult<Boolean> invoke(@ParameterName("input") String input, @ParameterName("pattern") String pattern) {
+    public FEELFnResult<Boolean> FEELFnResult(@ParameterName("input") String input, @ParameterName("pattern") String pattern) {
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/MatchesFunction.java:39]: @bncriju Wrong method name

**Review by yesamer — CHANGES_REQUESTED** [review]
@bncriju Thank you for your PR! After @gitgabrio and I analyzed your changes, we discovered more parts of the code that require to be improved. So, we agreed to open a new PR based on your PR to apply the adjustments.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/util/XQueryImplUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Test
+    void executeMatchesFunctionTest() {
+        String input = "test";
+        String pattern = "^test";
+        String flags = "i";
+        boolean retrieved = XQueryImplUtil.executeMatchesFunction(input, pattern,
+                flags);
+        boolean expected = true;
+        assertThat(retrieved).isNotNull().isEqualTo(expected);
```
Comment (pibizza) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/util/XQueryImplUtilTest.java:36]: One general comment. Do not reuse variables. Either separate the tests or make a parametric test. a test should be quite simple and very easy to read.

**Comment by yesamer** [comment]
Superseded by https://github.com/apache/incubator-kie-drools/pull/6085

**Review by pibizza — CHANGES_REQUESTED** [review]
Make the tests without reused variables. either separate tests or make them just parametric.

**Pull request closed without merging**